File: training/bkdnoj/dcl24demow/D.py
n = int(input())
target = input()
strings = [input().strip() for _ in range(n)]

target_length = len(target)
min_steps = [float('inf')] * (target_length + 1)
min_steps[0] = 0

for i in range(1, target_length + 1):
    for string in strings:
        if i >= len(string) and target[i - len(string):i] == string:
            min_steps[i] = min(min_steps[i], min_steps[i - len(string)] + 1)

result = min_steps[target_length]
print(result)

File: training/bkdnoj/dcl24demow/D.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    cin >> n;
    string target;
    cin >> target;
    vector<string> strings(n);
    for (int i = 0; i < n; ++i) {
        cin >> strings[i];
    }

    int m = target.size();
    vector<int> dp(m + 1, INT_MAX);
    dp[0] = 0;

    for (int i = 0; i < m; ++i) {
        if (dp[i] == INT_MAX) continue;
        for (int j = 0; j < n; ++j) {
            int len = strings[j].size();
            if (i + len <= m && target.substr(i, len) == strings[j]) {
                dp[i + len] = min(dp[i + len], dp[i] + 1);
            } else {
                for (int k = 1; k <= len && i + k <= m; ++k) {
                    if (target.substr(i, k) == strings[j].substr(0, k)) {
                        dp[i + k] = min(dp[i + k], dp[i] + 1);
                    }
                }
            }
        }
    }

    int result = dp[m] == INT_MAX ? -1 : dp[m];
    cout << result << endl;
    return 0;
}
